Refactored error handling in GameActions, created appropiate endpoint

Duckierstone42 opened the pull request:
### So far this is what I have done:

-Removed error handling logic in Game Actions (would just propagate upward anyway)

-Implemented a new Game Zod type that differentiates between storing themes and tags as objects instead of ID's

-Created the appropriate endpoint and action for filtering games

-Enable filtering by game build when Nathan's ticket gets merged.

### I still have to do  the following:

Fix problems with JEST testing, I was able to set up the whole testing environment but I'm having issues overwriting connectDB to use actions inside of it.

Duckierstone42 committed:
Refactored error handling in GameActions, created appropiate endpoint and action as well as zod types for game filtering in the backend, with pagination

Duckierstone42 committed:
Fixed issue with filtering by gameContent

katsukixyz commented on src/utils/types/index.ts:
```diff
@@ -35,6 +61,67 @@ export const gameSchema = z.object({
+    });
+    return z.NEVER;
+  }
+};
```
Array fields will be passed like this: `/api/games?themes="theme1"&themes="theme2"&themes="theme3"`, not in a serialized JSON format. This means that you can expect the value of the field to already be populated as an array of strings instead of needing to parse them

katsukixyz commented on src/utils/types/index.ts:
```diff
@@ -35,6 +61,67 @@ export const gameSchema = z.object({
+    .string()
+    .transform(convert_JSON)
+    .pipe(z.array(z.string().refine(validGameContent)))
+    .optional(),
```
For `gameBuilds` and `gameContent`, consider using a zod enum array instead - this will prevent you from having to write custom `.refine()` logic

katsukixyz commented on src/utils/types/index.ts:
```diff
@@ -35,6 +61,67 @@ export const gameSchema = z.object({
+};
+//For validating game query in the GET endpoint
+//Be aware that zod.transform alters the data during parsing.
+export const GetGameQuerySchema = z.object({
```
Nit: declare this next to the API route so it's easier to reference (in `/api/games/index.ts`)

katsukixyz commented on src/utils/types/index.ts:
```diff
@@ -9,6 +10,29 @@ const verifyObjectId = (value: string) => {
+    "videoTrailer",
+  ];
+  return validContent.includes(value);
+};
```
See previous comment - I think we can safely rid of this

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
+) {
+  await connectMongoDB();
+  //Want to be able to filter by specific tags,game_builds, etc.
+  const RESULTS_PER_PAGE = 6;
```
Nit: define this outside the method

katsukixyz commented on src/server/db/models/GameModel.ts:
Upon reading [these](https://mongoosejs.com/docs/typescript/populate.html) docs from mongoose, I think we shouldn't be typing the ref fields as the actual objects they represent. The actual documents that the refs link to only show up in the result of a `find()` if the `.populate()` step is used. Notice how in the first code snippet, child is typed as `Types.ObjectId`. We should adopt this pattern as well -- in our case, with arrays of `Types.ObjectId` ([this is different from `ObjectId`](https://arc.net/l/quote/vtsfsgbp)).

katsukixyz commented on src/utils/types/index.ts:
```diff
@@ -35,6 +61,67 @@ export const gameSchema = z.object({
   videoTrailer: z.string().url().optional(),
 });
 
+export const gameSchemaAPI = z.object({
+  //Make sure to modify gameSchema endpoints, as well as editGameSchema I suppose.
+  name: z.string().min(3).max(50),
+  themes: z.array(z.string().refine(verifyObjectId)).optional(),
+  tags: z.array(z.string().refine(verifyObjectId)).optional(),
+  description: z.string(),
+  game: z.string().url(),
+  lesson: z.string().url().optional(),
+  parentingGuide: z.string().url().optional(),
+  answerKey: z.string().url().optional(),
+  videoTrailer: z.string().url().optional(),
+});
```
What is the purpose of this schema?

Duckierstone42 replied:
The original gameSchema was altered and then used for a different purpose, so I created a new version called gameSchemaAPI that I'm using specifically for parsing at the game endpoints.

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
-      throw new GenericUserErrorException("Game with given ID does not exist.");
+    //Ensure that we found all tags and no extras
+    if (query.accessibility.length !== accessibility_tags.length) {
+      throw new GenericUserErrorException(
```
Define an exception for `Game` -- this one is for `User` routes

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
+  //Currently ignoring filtering by build, as that hasn't yet been implemented.
+
+  //Match by name with regex
+  if (query.name && query.name !== "") {
+    const reg_string = new RegExp(query.name, "i");
+    filters.name = { $regex: reg_string };
+  }
+  //Match by gameContent
+  if (query.gameContent && "answerKey" in query.gameContent) {
+    filters.answerKey = { $exists: true };
+  }
+  if (query.gameContent && "parentingGuide" in query.gameContent) {
+    filters.parentingGuide = { $exists: true };
+  }
+  if (query.gameContent && "lessonPlan" in query.gameContent) {
+    filters.lessonPlan = { $exists: true };
+  }
+  if (query.gameContent && "videoTrailer" in query.gameContent) {
+    filters.videoTrailer = { $exists: true };
+  }
```
Define a variable that maps each field in `query` to the filter object it needs to have, or a function that provides the object it needs to have:
```js
function gameContentHandler(gameContent: Pick<z.infer<typeof GetGameQuerySchema>, "gameContent">) {
  const uniqueGameContent = new Set(gameContent);
  const filterObj = Array.from(uniqueGameContent).reduce((acc, cur) => {
    return {
      ...acc,
      [cur]: { $exists: true }
    }
  }, {});
  return filterObj;
}
```


```js
{
  name: { $regex: reg_string },
  themes: { $all: foundThemes }
  gameContent: gameContentHandler(query.gameContent),
}
```
and then you can iterate over all fields in `query` that are defined and build the filter object using `Array.reduce`.

katsukixyz replied:
**Side note that you might find helpful** - the `in` operator in JS does not look at the values, but rather the properties of what it is applied on. In the case of arrays, you can use `in` to see if an index exists or if a property like `"length"` exists on the object, but this doesn't check the value itself, so your conditions would always return false. This may have introduced a bug that you may have noticed upon doing your unit tests. See [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Expressions_and_Operators#in)

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
-  } catch (e) {
-    throw e;
+    //Filter used to ensure object is not null, then map converts it to an array of object ids.
+    total_tags = total_tags.concat(
```
Nit: use camel case

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
-    throw e;
+    //Filter used to ensure object is not null, then map converts it to an array of object ids.
+    total_tags = total_tags.concat(
+      accessibility_tags.filter((tag) => tag).map((tag) => tag.id),
```
Is a null check for each tag necessary here?

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
-    if (games == null) {
-      return [];
+  if (query.tags) {
+    const normal_tags = await TagModel.find({ name: { $in: query.tags } });
```
Check for `type: "custom"` here

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
-  } catch (e) {
-    throw e;
-  }
+  const game = await GameModel.findById(id).populate("themes").populate("tags");
```
After changing the schema as discussed earlier, pass in a type into `populate<...>()`

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
+      }
+    });
+  }
+  if (data && data.tags) {
+    const tagPromises = data.tags.map((tag) => TagModel.findById(tag));
+    const tagResults = await Promise.all(tagPromises);
+    tagResults.forEach((result, index) => {
+      if (!result) {
+        throw new GenericUserErrorException(
+          `ObjectID ${data.tags![index]} is not a  present tag.`,
+        ); //Using non-null assertion, as if condition should ensure data.tags is non-null
+      }
+    });
+  }
```
Consider using the `$in` operator here too

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
+  //Ensure every ObjectID actually represents a Document
+  //Ah yes
+  if (data && data.themes) {
+    const themePromises = data.themes.map((theme) =>
+      ThemeModel.findById(theme),
+    );
+    const themeResults = await Promise.all(themePromises);
+    themeResults.forEach((result, index) => {
+      if (!result) {
+        throw new GenericUserErrorException(
+          `ObjectID ${data.themes![index]} not a present theme.`,
+        ); //Using non-null assertion, as if condition should ensure data.tags is non-null
+      }
+    });
```
Here as well

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -1,62 +1,55 @@
-  } catch (e) {
-    throw e;
+  if (data && data.themes) {
+    const themePromises = data.themes.map((theme) =>
+      ThemeModel.findById(theme),
+    );
+    const themeResults = await Promise.all(themePromises);
+    themeResults.forEach((result, index) => {
+      if (!result) {
+        throw new GenericUserErrorException(
+          `ObjectID ${data.themes![index]} not a present theme.`,
+        ); //Using non-null assertion, as if condition should ensure data.tags is non-null
```
See above

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -1,62 +1,55 @@
-  } catch (e) {
-    throw e;
+  if (data && data.tags) {
+    const tagPromises = data.tags.map((tag) => TagModel.findById(tag));
+    const tagResults = await Promise.all(tagPromises);
+    tagResults.forEach((result, index) => {
+      if (!result) {
+        throw new GenericUserErrorException(
+          `ObjectID ${data.tags![index]} is not a  present tag.`,
+        ); //Using non-null assertion, as if condition should ensure data.tags is non-null
+      }
+    });
```
See above

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
-    return games;
-  } catch (e) {
-    throw e;
+    total_tags = total_tags.concat(
```
You can instead do `const totalTags = [...accessibilityTags, ...normalTags]` for readability

katsukixyz commented on src/utils/types/index.ts:
```diff
@@ -35,6 +61,67 @@ export const gameSchema = z.object({
+    .transform(convert_JSON)
+    .pipe(z.array(z.string().refine(validGameContent)))
+    .optional(),
+  page: z.string().transform(convert_INT).pipe(z.number().gte(1)).optional(),
```
Make this required - we can assume the frontend will always provide a page

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -68,70 +61,131 @@ export async function editGame(allData: nextEditGame) {
+      .limit(RESULTS_PER_PAGE);
+  } else {
+    games = await GameModel.find(filters);
+  }
```
`query.page` will always be defined, so you can keep the skip and limit steps.

katsukixyz reviewed (commented):
Good start!! Left some comments that hopefully will help you out 😄

Duckierstone42 committed:
Implemented all of Katsuki's comments

Duckierstone42 committed:
Merge branch 'main' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

Duckierstone42 committed:
Removed duplicate IGame import

Duckierstone42 committed:
Added filtering based on gameBuild

Duckierstone42 committed:
Merge branch 'main' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

Duckierstone42 committed:
Fixed issue with package.json

Duckierstone42 committed:
Removd crlf and replaced with lf to avoid issues with prettier

katsukixyz commented on src/pages/api/games/[id]/index.ts:
```diff
@@ -44,7 +44,7 @@ export default async function handler(
       case "PUT":
         //If no ID in query or not valid
-        if (!gameId || !mongoose.Types.ObjectId.isValid(gameId)) {
+        if (!gameId || !ObjectId.isValid(gameId)) {
```
Use `isValidObjectId` instead: see [here](https://arc.net/l/quote/gyotuvoz)

katsukixyz commented on src/pages/api/games/index.ts:
```diff
@@ -1,17 +1,82 @@
+import { z, RefinementCtx } from "zod";
+//Ensure its an integer
+// GameContentEnum
+const GameContentEnum = {
+  answerKey: "answerKey",
+  parentingGuide: "parentingGuide",
+  lessonPlan: "lesson",
+  videoTrailer: "videoTrailer",
+};
```
Declare this as an actual enum for type safety: see TS [docs](https://www.typescriptlang.org/docs/handbook/enums.html#string-enums)

katsukixyz commented on src/pages/api/games/index.ts:
```diff
@@ -1,17 +1,82 @@
+};
+
+// GameBuildsEnum
+const GameBuildsEnum = {
+  amazon: "amazon",
+  android: "android",
+  appstore: "appstore",
+  linux: "linux",
+  mac: "mac",
+  webgl: "webgl",
+  windows: "windows",
+};
```
Here too

katsukixyz commented on src/pages/api/games/index.ts:
```diff
@@ -1,17 +1,82 @@
+    .array(z.string())
+    .or(z.string().transform(putSingleStringInArray))
+    .optional(),
+  accessibility: z
```
Would it be alright to assume that it's just given as an array of strings?

katsukixyz commented on src/pages/api/games/index.ts:
```diff
@@ -1,17 +1,82 @@
-        const games = await getAllGames();
+        //Validate req.query
+        console.log(req.query);
+        const query = GetGameQuerySchema.parse(req.query); //JSON.parse not necessary
```
Use `safeParse` to handle the invalid case more directly

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -72,70 +72,124 @@ export async function editGame(allData: nextEditGame) {
+  let filters: FilterQuery<IGame> = {}; //There should be a better way to build the query gradually
+  if (query.theme) {
+    //Find corresponding themeObject in mongodb
+    const found_theme: (ITheme & { id: ObjectId }) | null =
```
You shouldn't need to assert this type

Nit: use camelCase

katsukixyz replied:
Anytime you reference ObjectId as a **type**, use `Types.ObjectId` instead of just `ObjectId`. Read [here](https://mongoosejs.com/docs/schematypes.html#:~:text=A%20SchemaType%20is%20different%20from%20a%20type.%20In%20other%20words%2C%20mongoose.ObjectId%20!%3D%3D%20mongoose.Types.ObjectId)

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -72,70 +72,124 @@ export async function editGame(allData: nextEditGame) {
       });
+    if (!found_theme) {
+      //No theme was found
+      throw new GenericGameErrorException(
```
Consider making an exception for this case - generic should be left to errors you don't know at compile time

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -72,70 +72,124 @@ export async function editGame(allData: nextEditGame) {
+      type: "accessibility",
     });
-    if (!result) {
-      throw new GenericUserErrorException("Game with given ID does not exist.");
+    //Ensure that we found all tags and no extras
+    if (query.accessibility.length !== accessibilityTags.length) {
+      throw new GenericGameErrorException(
+        "One or more of the accessibility tags do not exist.",
+      );
     }
-    return result;
-  } catch (e) {
-    throw e;
+    accessibility = accessibilityTags.map((tag) => tag.id);
   }
-}
+  let custom: ObjectId[] = [];
+  if (query.tags) {
+    const normalTags = await TagModel.find({
+      name: { $in: query.tags },
+      type: "custom",
+    });
+    if (query.tags.length !== normalTags.length) {
+      throw new GenericGameErrorException(
+        "One or more of the custom tags do not exist.",
+      );
+    }
+    custom = normalTags.map((tag) => tag.id);
+  }
```
Consider refactoring conditional checks for fields in `query` for readability, I think I left a [comment](https://github.com/GTBitsOfGood/jennifer-anns-group/pull/39#discussion_r1501647268) about this last time, applies to all of these conditional checks

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -72,70 +72,124 @@ export async function editGame(allData: nextEditGame) {
-      return [];
+  //Match by name with regex
+  if (query.name && query.name !== "") {
+    const reg_string = new RegExp(query.name, "i");
```
This should match partial matches as well - not just exact matches, use `*` to match 0 or more characters in both directions

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -72,70 +72,124 @@ export async function editGame(allData: nextEditGame) {
-  } catch (e) {
-    throw e;
-  }
+  const game = await GameModel.findById(id).populate<IGame>("themes tags");
```
The type passed via `.populate()` should represent the types of the objects that are populated:

```js
.populate<ITheme>("themes").populate<ITag>("tags")
```

katsukixyz commented on src/server/db/models/GameModel.ts:
```diff
@@ -16,12 +21,11 @@ const GameSchema = new Schema<IGame>({
   themes: {
     type: [Schema.Types.ObjectId],
-    ref: "Theme",
+    default: [],
```
Keep the ref to the `Theme` document, this is necessary for other backend routes

katsukixyz commented on src/utils/types/index.ts:
```diff
@@ -1,3 +1,4 @@
+import { setDefaultAutoSelectFamily } from "net";
```
remove this import

katsukixyz commented on src/utils/exceptions.ts:
```diff
@@ -23,6 +23,11 @@ export class GenericUserErrorException extends Error {
     super(message);
   }
 }
+export class GenericGameErrorException extends Error {
+  constructor(message = "User mistake in game endpoint") {
+    super(message);
+  }
+}
```
Make an abstract class for a game exception and let all the game exceptions extend it. This will make error handling easier

Duckierstone42 committed:
Implemented Katsuki's second round of comments

Duckierstone42 committed:
Merge branch 'main' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

Duckierstone42 committed:
Added basic testing structure with JEST, just need to add actual tests for getSelectedGames

Duckierstone42 committed:
Partial jest tests

03hchen commented on src/pages/api/games/index.ts:
```diff
@@ -23,9 +28,73 @@ export default async function handler(
+  android = "android",
+  appstore = "appstore",
+  linux = "linux",
+  mac = "max",
```
= "mac"

03hchen commented on src/server/db/actions/GameAction.ts:
```diff
@@ -79,54 +87,118 @@ interface nextEditGame {
-        ); //Using non-null assertion, as if condition should ensure data.tags is non-null
-      }
+  if (data && data.themes) {
+    const themeResults = await ThemeModel.find({ id: { $in: data.themes } });
```
This might've been a change I made in main that was merged in but I believe this should be `_id`? I think Lauren already made this change in her branch though

03hchen commented on src/utils/exceptions.ts:
```diff
@@ -0,0 +1,92 @@
+import { MongoServerError } from "mongodb";
```
hmm I believe the backend refactoring done a few days ago removed this file and moved all exceptions. Could you move each of these errors into their respective `src/utils/exceptions` folder (i think some are duplicates, so just make sure the imports are to the individual folders and not this file) and then remove this file?

03hchen commented on src/utils/types/index.ts:
```diff
@@ -51,6 +53,10 @@ export const gameSchema = z.object({
+//Since arrays from req.query are just strings, and need to be converted into arrays.
+
+//For validating game query in the GET endpoint
+//Be aware that zod.transform alters the data during parsing.
```
I'm kinda confused what these comments mean, can they be removed?

Duckierstone42 replied:
Yeah they can be removed, a lot of these comments are me just thinking through the code before writing it.

Duckierstone42 committed:
Refactored code to include Helen's comments

Refactoed getSelectedGames as previous version wasn't working with the updated gameBuilds in the gameModel schema. Now instead of find I am using aggregate, as I need to transform the data midway for extra filtering, then transform it back to its original state.

Unfortunately JEST testing is still not functional, still having issues overiding connectDB

Duckierstone42 committed:
Resolved merge conflict after pulling from main

Duckierstone42 committed:
Merge branch 'main' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

katsukixyz committed:
fix: mongodb connection, jest in-memory db

Duckierstone42 committed:
Merge branch 'ankith/game-gallery-backend' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

Duckierstone42 committed:
Added unit tests :(

Duckierstone42 committed:
Removed console logs

Duckierstone42 committed:
Made it compare objects directly in serialized format

katsukixyz committed:
chore: stub unit tests, examples

Duckierstone42 committed:
Added all tests

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+  });
+
+  beforeEach(async () => {
+    generatedGames = randomGames(NUM_GAMES);
```
Move this outside of beforeEach

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+          game.tags = [...game.tags!, tag._id];
+        }
+      });
+    });
```
Define this outside as well

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+        if (gameIDs.includes(game._id)) {
+          game.themes ?? [];
+          //The nullish coalescing operators ensures game.themes is defined
+          game.themes = [...game.themes!, themeInput._id];
```
`game.themes` is optional as specified in `GameModel`, so we shouldn't set it as an empty array if there are no themes assigned

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+        if (gameIDs.includes(game._id)) {
+          game.tags ?? [];
+          //The nullish coalescing operators ensures game.themes is defined
+          game.tags = [...game.tags!, tag._id];
```
See above

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+      //Ensure that the length of the in-group is the same as the one of the in group.
+      const tag = faker.helpers.arrayElement(tagInputs);
+      //Loop through all pages to make a list of all games with that tag
+      let query: GameQuery = {
+        page: 1,
+      };
+      //Filter generated gams
+      if (tag.type == "custom") {
+        query.tags = [tag.name];
+      } else {
+        query.accessibility = [tag.name];
+      }
+
+      let results = await getSelectedGames(query);
+      let games = results.games;
+      const numPages = Math.ceil(results.count / RESULTS_PER_PAGE) - 1; //-1 Since one page has already ben traversed.
+      for (let page = 2; page < 2 + numPages; page++) {
+        query.page = page;
+        const result = await getSelectedGames(query);
+        games = [...games, ...result.games];
+      }
```
It's a bit counterintuitive to write the test this way - if you're using a loop to iterate through each page, you shouldn't make the call first then enter the loop like you're doing here. Consider rewriting this so that you have 1 loop doing all the functionality you want to accomplish

Duckierstone42 replied:
I'm currently using the first call outside the loop to determine the number of pages to use. Do you know how I might be able to get the number of pages without using that call outside the loop and avoiding errors?

katsukixyz replied:
You don't necessarily have to make that call since this unit test is **specifically** testing the exact results of `tags`. This means that you can assume your action will correctly give you the correct number of pages, so you can derive this from `Math.ceil(NUM_GAMES / RESULTS_PER_PAGE)`.

Duckierstone42 replied:
Removed both preliminary calls and simply used tag.games.length and the appropiate one for themes.

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+        query.page = page;
+        const result = await getSelectedGames(query);
+        games = [...games, ...result.games];
+      }
```
See above

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+      theme.games.sort();
+      expect(gameIds).toStrictEqual(theme.games);
+    });
+    test("[name] regex case insensitivity in-exact match: expect success", async () => {
```
You should define a short substring with a single character for this test, instead of randomly capitalizing a substring. This is because you want to be testing for multiple occurrences instead of just one game.

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+          page: page,
+        });
+        games = [...games, ...result.games];
+      }
```
See above

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+        accessibilityTags.every((tag) => game.tags?.includes(tag)),
+      ), //TODO: rework
+    theme: (games, themeName, _) =>
+      games.filter((game) => game.themes?.includes(themeName)),
```
You'll likely need to rework this because `games.themes` is an array of ObjectIds, not names of themes

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+      ),
+    tags: (games, customTags, _) =>
+      games.filter((game) =>
+        customTags.every((tag) => game.tags?.includes(tag)),
```
Same here - `game.tags` is an array of ObjectIds so youll need to scan the themes and tags you generate to find the id for them

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+      );
+      console.log("Filtered games:", key, filteredGames);
+    }
+    return filteredGames;
```
Just noticed this now: I told you about using `page` to as the final step here that I forgot, but we also need to sort everything by the ObjectId before that. The random generators for the games **should** produce ObjectIds in a predictable way (ascending), but let's be sure in the implementation for getting the expected results.

Duckierstone42 replied:
Currently sorting by name instead (as that is what we intended to do in the future)

katsukixyz commented on src/server/db/actions/GameAction.ts:
```diff
@@ -14,6 +16,16 @@ import {
 
+export const RESULTS_PER_PAGE = 7;
+
+export const buildConverter: { [key: string]: string } = {
```
Use `Record<AppType, string>` instead.

katsukixyz reviewed (commented):
Left some comments! Also a side-note and a tiny nit: the comments in your code can make it slightly harder for a reviewer to read your code. Consider taking them out of the commit when you push your code (if it helps you, you can stash your changes before you commit and then pop it so that you get the comments back after you push).

katsukixyz commented on src/__tests__/server/db/actions/GameAction.unit.test.ts:
```diff
@@ -0,0 +1,355 @@
+        games: ExtendId<IGame>[];
+        count: number;
+      };
+      try {
+        actual = await getSelectedGames(query);
+      } catch (e) {
+        if (e instanceof GameNotFoundException) {
+          actual = { games: [], count: 0 }; //Accounts for the case when we grab none of the games via the test.
+        } else {
+          throw e;
+        }
+      }
```
Oops, forgot to leave a comment on this - avoid using a try-catch block in a unit test. You should aim to be use matchers using `expect` whenever possible.

Duckierstone42 committed:
Partially added comments

03hchen committed:
uhhh

Duckierstone42 committed:
current changes

03hchen committed:
Merge branch 'ankith/game-gallery-backend' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

Duckierstone42 committed:
Temps tuff

03hchen committed:
Merge branch 'ankith/game-gallery-backend' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

Duckierstone42 committed:
Added a bunch of refactoring

03hchen committed:
Merge branch 'ankith/game-gallery-backend' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

Duckierstone42 committed:
Another comit

03hchen committed:
Merge branch 'ankith/game-gallery-backend' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

Duckierstone42 committed:
Removed two typescript errors

Duckierstone42 committed:
Removed comments

03hchen committed:
Merge branch 'ankith/game-gallery-backend' of https://github.com/GTBitsOfGood/jennifer-anns-group into ankith/game-gallery-backend

katsukixyz committed:
fix: typescript error

katsukixyz committed:
fix: gamebuilds expected and actual test

Duckierstone42 committed:
Refacted getSelectedGames endpoint to mimic how the tests are setup

Duckierstone42 committed:
Remove any type in aggregate with project stage

03hchen committed:
minor refactoring

03hchen committed:
Merge branch 'main' into ankith/game-gallery-backend

03hchen committed:
package json stuff

03hchen committed:
final changes
